packages/dds/tree/src/simple-tree/api/schemaFactory.ts

```diff
@@ -890,3 +896,72 @@ export function markSchemaMostDerived(schema: TreeNodeSchema): void {
 
 	(schema as typeof TreeNodeValid & TreeNodeSchema).markMostDerived();
 }
+
+/**
+ * Binds metadata to a node schema.
+ * @remarks Accomplishes this by creating a new subclass with the specified metadata, strongly typed.
+ *
+ * @param nodeSchema - The node schema to bind metadata to.
+ * @param metadata - The metadata to bind to the node schema.
+ *
+ * @example
+ *
+ * ```typescript
+ * class Point extends withMetadata(
+ * 	schemaFactory.object("point", { x: schemaFactory.number, y: schemaFactory.number }),
+ * 	{
+ * 		description: "A point in 2D space",
+ * 		custom: {
+ * 			... // Your custom metadata properties here
+ * 		},
+ * 	}
+ * ) {}
+ * ```
+ *
+ * @alpha
+ */
+export function withMetadata<
+	TNewCustomMetadata = unknown,
+	TName extends string = string,
```

Comment:
@CraigMacomber a sanity check on the type defaults here would be helpful. These were taken from `TreeNodeSchemaClass`'s defaults, but it isn't clear to me if those are the best match for the public SchemaFactory APIs.

Reply:
Generally default types for functions are useful when there is a type before them which you might want to manually specify, then use the defaults for the rest (sadly you can't specify one and infer the rest). THis fact makes these defaults more harmful than useful, since it looks like you could pass in the TNewCustomMetadata type as the type parameter here and have it work, but doing that will erase all the details of the schema.


Additionally, if the call being passed is has extra members (ex: custom ones, or things from ObjectSchema if we make it public) those will be lost due to how this is written, but should be preserved because of how subclassing works.

THis makes me thing it should take in a `TSchema extends TreeNodeSchemaClass` with no default and return `TSchema & {metadata stuff}`.

Sadly, this isn't correct either, since that won't work if overriding existing metadata (since the types of the field would be intersected I think, which isn't correct or sound).

Annoyingly you can't use Omit to remove the metadata field from TSchema to fix this since IIRC Omit loses protected and constructor members. Thus you could have to omit the metadata from the schema, add in the new type, then restore that its a constructor manually, I think.

Reply:
That is all getting very complicated, and seems particularly brittle to future changes. All of this makes me think that it would be better to go back to the original approach of adding an `options` property bag to node schema factory methods.